src/acl.c

```diff
@@ -1507,6 +1507,10 @@ static int ACLSelectorCheckKey(aclSelector *selector, const char *key, int keyle
     listRewind(selector->patterns,&li);
 
     int key_flags = 0;
+    if (keyspec_flags & CMD_KEY_RO) key_flags |= ACL_READ_PERMISSION;
+    if (keyspec_flags & CMD_KEY_RW) key_flags |= ACL_ALL_PERMISSION;
+    if (keyspec_flags & CMD_KEY_OW) key_flags |= ACL_WRITE_PERMISSION;
+    if (keyspec_flags & CMD_KEY_RM) key_flags |= ACL_WRITE_PERMISSION;
```

Comment:
according to our definition of the key-spec flags, we defined that certain parts of the key access are not considered read (e.g. LFU, cardinality, length, existence).
we did that in order to better support the write-only requirement (i.e don't consider the return value from LPUSH a read operation).
if we follow that design, then LLEN and EXIST do not require user read permission, but you are right that it seems odd that users can do these things on keys that the ACL didn't define as accessible in any way.
and also it kinda breaks existing use cases.

my suggestion is to add a 3rd ACL flag, ACL_PRESENCE_PERMISSION (p.s. i think it should be ACL_PERMISSION_PRESENCE, i.e. same prefix on all 3 flags).
then, we should set this flag in selectors with either READ or WRITE.
and set that flag here on RO and RW.

Reply:
actually, this change goes too deep into the ACL and it's DSL. i think we can solve it with a small change in ACLSelectorCheckKey.
i'm looking into it.

Reply:
Ohh, scrap what i written before. i didn't understand your complaint.